pinot-hadoop/src/test/java/com/linkedin/pinot/hadoop/io/PinotOutputFormatTest.java

```diff
@@ -0,0 +1,224 @@
+        PinotOutputFormat.setTempSegmentDir(job, workingTempDir.getAbsolutePath());
+
+        Schema schema = Schema.fromString(getSchema());
+        PinotOutputFormat.setSchema(job, schema);
+        PinotOutputFormat.setNumThreads(job, 10);
+        mockTaskAttemptContext();
+    }
+
+    private void mockTaskAttemptContext() {
+        TaskAttemptID fakeTaskId = new TaskAttemptID(new TaskID("foo_task", 123, TaskType.REDUCE, 2), 2);
+        when(fakeTaskAttemptContext.getTaskAttemptID())
+                .thenReturn(fakeTaskId);
+        when(fakeTaskAttemptContext.getConfiguration()).thenReturn(job.getConfiguration());
+    }
+
+
+    @Test
+    public void verifyStarIndex() throws Exception {
+        int days = 2000;
+        int sal = 20;
+        PinotOutputFormat.setEnableStarTreeIndex(job, true);
+        RecordWriter<Object, byte[]> writer = outputFormat.getRecordWriter(fakeTaskAttemptContext);
+        ObjectMapper mapper = new ObjectMapper();
+        Map<Integer, Emp> inputMap = new HashMap<>();
+        for (int i = 0; i < 100; i++) {
+            String name = "name " + i;
+            Emp e = new Emp(i, name, days + i, sal + i);
+            byte[] b = mapper.writeValueAsBytes(e);
+            writer.write(null, b);
+            inputMap.put(i, e);
+        }
+        writer.close(fakeTaskAttemptContext);
+
+        String segmentTarPath = "_temporary/0/_temporary/attempt_foo_task_0123_r_000002_2/part-r-00002/segmentTar";
+        File segmentTarOutput = new File(outputTempDir, segmentTarPath);
+        File untarOutput = Files.createTempDirectory(PinotOutputFormatTest.class.getName() + "_segmentUnTar").toFile();
+        for (File tarFile : segmentTarOutput.listFiles()) {
+            TarGzCompressionUtils.unTar(tarFile, untarOutput);
+        }
+
+        File outputDir = new File(untarOutput, PinotOutputFormat.getSegmentName(fakeTaskAttemptContext));
+        RecordReader recordReader = new PinotSegmentRecordReader(outputDir, Schema.fromString(getSchema()));
+        GenericRow row = new GenericRow();
+        Map<Integer, GenericRow> resultMap = new HashMap<>();
+        while (recordReader.hasNext()) {
+            row = recordReader.next(row);
+            resultMap.put((Integer) row.getValue("id"), row);
+            row.clear();
+        }
+
+        Assert.assertEquals(resultMap.size(), inputMap.size());
+        // more data validation.
+    }
+
+    @Test
+    public void verifyRawIndex() throws Exception {
+        int days = 2000;
+        int sal = 20;
+        PinotOutputFormat.setEnableStarTreeIndex(job, false);
+        RecordWriter<Object, byte[]> writer = outputFormat.getRecordWriter(fakeTaskAttemptContext);
+        ObjectMapper mapper = new ObjectMapper();
+        Map<Integer, Emp> inputMap = new HashMap<>();
+        for (int i = 0; i < 100; i++) {
+            String name = "name " + i;
+            Emp e = new Emp(i, name, days + i, sal + i);
+            byte[] b = mapper.writeValueAsBytes(e);
```

Comment:
This test case failing from the second record, throwing
`java.lang.IndexOutOfBoundsException
	at java.nio.Buffer.checkIndex(Buffer.java:546)
	at java.nio.DirectByteBuffer.getInt(DirectByteBuffer.java:685)
	at com.linkedin.pinot.core.segment.memory.PinotByteBuffer.getInt(PinotByteBuffer.java:195)
	at com.linkedin.pinot.core.io.util.FixedByteValueReaderWriter.getInt(FixedByteValueReaderWriter.java:37)
	at com.linkedin.pinot.core.io.reader.impl.v1.SortedIndexReader.getInt(SortedIndexReader.java:85)
	at com.linkedin.pinot.core.io.reader.impl.v1.SortedIndexReader.getInt(SortedIndexReader.java:27)
	at com.linkedin.pinot.core.data.readers.PinotSegmentColumnReader.readInt(PinotSegmentColumnReader.java:51)
	at com.linkedin.pinot.core.data.readers.PinotSegmentRecordReader.next(PinotSegmentRecordReader.java:107)
	at com.linkedin.pinot.hadoop.io.PinotOutputFormatTest.verifyRawIndex(PinotOutputFormatTest.java:153)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:76)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:673)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:846)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1170)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.runWorkers(TestRunner.java:1147)
	at org.testng.TestRunner.privateRun(TestRunner.java:749)
	at org.testng.TestRunner.run(TestRunner.java:600)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:317)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:312)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:274)
	at org.testng.SuiteRunner.run(SuiteRunner.java:223)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1039)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:964)
	at org.testng.TestNG.run(TestNG.java:900)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)
`